x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-DeltaAggregator.java.st

```diff
@@ -0,0 +1,235 @@
+            this.bigArrays = bigArrays;
+            this.breaker = breaker;
+            this.states = bigArrays.newObjectArray(1);
+        }
+
+        void ensureCapacity(int groupId) {
+            states = bigArrays.grow(states, groupId + 1);
+        }
+
+        void adjustBreaker(long bytes) {
+            breaker.addEstimateBytesAndMaybeBreak(bytes, "<<delta aggregation>>");
+            stateBytes += bytes;
+            assert stateBytes >= 0 : stateBytes;
+        }
+
+        void append(int groupId, long timestamp, $type$ value) {
+            var state = states.get(groupId);
+            if (state == null) {
+                state = new $Type$DeltaState(timestamp, value);
+                states.set(groupId, state);
+                adjustBreaker(state.bytesUsed());
+            } else {
+                if (timestamp >= state.lastTimestamp) {
+                    state.lastTimestamp = timestamp;
+                    state.lastValue = value;
+                    state.valuesSeen++;
+                } else if (timestamp <= state.firstTimestamp) {
+                    state.firstTimestamp = timestamp;
+                    state.firstValue = value;
+                    state.valuesSeen++;
+                } // else: ignore, too old
+            }
+        }
+
+        void combine(int groupId, LongBlock timestamps, $Type$Block values, int otherPosition) {
+            final int valueCount = timestamps.getValueCount(otherPosition);
+            if (valueCount == 0) {
+                return;
+            }
+            final int firstTs = timestamps.getFirstValueIndex(otherPosition);
+            final int firstIndex = values.getFirstValueIndex(otherPosition);
+            ensureCapacity(groupId);
+            append(groupId, timestamps.getLong(firstTs), values.get$Type$(firstIndex));
+            if (valueCount > 1) {
+                ensureCapacity(groupId);
+                append(groupId, timestamps.getLong(firstTs + 1), values.get$Type$(firstIndex + 1));
+            }
+        }
+
+        @Override
+        public long ramBytesUsed() {
+            return states.ramBytesUsed() + stateBytes;
+        }
+
+        @Override
+        public void close() {
+            Releasables.close(states, () -> adjustBreaker(-stateBytes));
+        }
+
+        @Override
+        public void toIntermediate(Block[] blocks, int offset, IntVector selected, DriverContext driverContext) {
+            assert blocks.length >= offset + 2 : "blocks=" + blocks.length + ",offset=" + offset;
+            final BlockFactory blockFactory = driverContext.blockFactory();
+            final int positionCount = selected.getPositionCount();
+            try (
+                LongBlock.Builder timestamps = blockFactory.newLongBlockBuilder(positionCount * 2);
+                $Type$Block.Builder values = blockFactory.new$Type$BlockBuilder(positionCount * 2);
+            ) {
+                for (int i = 0; i < positionCount; i++) {
```

Comment:
I think we need to serialize `valuesSeen` so that we can apply extrapolation in the final?